### PR TITLE
[WIP] Conversion of spells to magic

### PR DESCRIPTION
**Pull request by chaosvolt**
Still very much a work in progress, as I'm still figuring out general balance so far, doing the actual spell conversions, and still discovering what I can do currently vs. what I can't.

For now, I've partially made conversions of the 15 basic spells (the ones available via monsterparts), with a general balance set up for now. Currently uses a maximum of 10 spell levels, and difficulty taken directly from the mutation  point cost (which is used informally as a rough estimate of spell strength), and mana cost calculated from difficulty (25 base cost, plus 25 per difficulty). Idea is they'll scale spell level up relatively quickly but won't shake off spell failure chances completely without above-average intelligence or some skill.

I'm still thinking about just what I'd do do differentiate Paragon of The Veil spells from normal spells. I would likely preserve the current logic of the overall cost being more efficient, might also give them a higher max level with a more prolonged ramp-up of effects, lower difficulty may or may not be added (higher max level means spell failure will cease to be an issue when advanced enough).

On the plus side, eliminating magic signs as items whenever possible will allow me to finally set aside consistent naming, so I can label mundane spells Magic Signs and Paragon of The Veil ones Arcane Blessings.

Current list of missing features evident:
* The big one is that the hardcode forces Spellcraft to be the relevant skill used for learning and casting spells, rather than Arcana. Given this feature will eventually have uses beyond magic, presumably being able to set what skill is used is part of future plans.
* I'm still torn between using mana vs. fatigue, but if I did want to continue using fatigue it's not really an option right now for spell resources.
* Overgrowth and any other active artifact effects that can't be replicated by spells in some way will simply not work right now at all. `spawn_item` isn't really ideal for this unless it was made into basically just a straight "casting it spawns the magic sign" thing.
* Related: highly-specialized spells like Overgrowth will basically have no way whatsoever to make spell levels relevant given the effect is so specific and wrapped up in hardcode. Even if I could abuse `spawn_item` to place a source of arti-effects at a distance, `AEA_PLANTS` still triggers the effect at the player's location, not where the item is actually used. Presumably a future JSONized way to convert terrain to mimic Overgrowth's behavior would be a better fit, and would likely allow for AoE to factor into the balance too.
* Can't conjure fields with attacks, currently relevant to Acid Spray and Shockstorm (the two offensive spellcraft spells). Worse, I dunno how stuff like handling more than one type of field will work, which might have some specialist uses down the line.
* Will want to wait to implement obtaining the spells until after gaining spells from mutations as a thing. I could make spellcraft scrolls basically using current learn items, but this method precludes being able to supersede normal spells with Paragon of The Veil versions (currently takes advantage of the mutations being able to supersede each other).
* Some of these spells will almost certainly need the ability to define multiple effects in some way. Wave of Destruction for example can likely do most of its effects as-is (pure damage and stunning of monsters), but separate side effects inflicted on the player and terrain destruction likely would be different effects entirely.
* Presumably a `use_item` spell effect would be handy for some things too?

Remainder on to-list:
* Convert rest of the 15 basic spells to this system to figure out what else may need to be done.
* Dabble in converting magic seals to use the pending "wand" system, will have to see if I can set them to be consumable or otherwise deplete on use, or else I'll have to retain the pre-existing fuckery where the item has to turn into an active version that activates and depletes itself a turn later (only way I've found to make a tool that vanishes or reverts on depletion is basically abusing the active state).
* Convert the rest of the spells and figure out what else might be a problem to overhaul.
* Figure out what I'm gonna do with Dragonblood's current balance. As-is, if I'm stuck using mana their mutations no longer provide that "the mutation paths benefit spellcasting in different ways" idea, because fatigue buffs will no longer impact how often you can cast spells.
* Whenever mutations can assign spells, I'll be able to actually properly provide hooks for gaining spells from restored ritual blade and sacramental heart, in addition to (possibly) doing the same with the 15 basic spells, since spell upgrades oughta still be a thing.
* Could set up dragonblood-exclusive spells and Veil-exclusive spells such that taking one path excludes benefiting from the other, but I dunno yet how it works beyond spellbooks giving a prompt to gain a trait and permanently lock yourself out of a given spell category, nor do I know how it'll interact with spells from mutations. As it is, I ALREADY have some ideas on how to make Dragonblood and Paragon of The Veil mutually exclusive...

### CODE REVIEW

**Comment by KorGgenT** [comment]
for that last point: current likely way is that gaining the mutation that's exclusive to other one will cancel the other mutation and make you forget all spells the other mutation would have given you.

**Comment by KorGgenT** [comment]
also, for your effects, see this PR https://github.com/CleverRaven/Cataclysm-DDA/pull/30845

**Comment by chaosvolt** [comment]
>for that last point: current likely way is that gaining the mutation that's exclusive to other one will cancel the other mutation and make you forget all spells the other mutation would have given you.

Makes sense, yeah. So basically handling it via mutation types or cancellation, like normal. Or did you mean also having the "spell type" system be checked for? Basically I don't yet know which method to make use of. I might use the latter for Dragonblood and Paragon of The Veil as mentioned, while leaving spellcraft spells

Problem with THAT is some spellcraft spells are also used by Dragonblood, so I'd have to basically split them into "spellcraft" and "Dragonblood" variants like I do with spells also available to Paragon of The Veil. Which will add more unique spells and mutations to a list that's currently sitting at *61* spell mutations.

> also, for your effects, see this PR

Nice. Assuming "BP-less" becomes the default behavior, then all the effects I've added so far need no change as they aren't BP-specific. When I convert fire and spells I can give them BP-targetting for on-fire and corroding.

**Comment by KorGgenT** [comment]
bp-less means not applied at all, because effects need a body part to latch onto

**Comment by chaosvolt** [comment]
Oh. That's...odd.

File: Arcana/spells.json [hunk 1]
```diff
@@ -0,0 +1,883 @@
...
+  },
+  {
+    "id": "arcana_fire",
+    "type": "SPELL",
+    "name": "Magic Sign: Conjure Flame",
+    "description": "Using this spell will conjure an eruption of flames at the location targeted.",
+    "effect": "target_attack",
+    "valid_targets": [ "hostile", "ground" ],
+    "flags": [ "NO_HANDS", "NO_LEGS", "VERBAL" ],
+    "effect_str": "onfire",
```
Comment (KorGgenT) [inline Arcana/spells.json]: heat damage already applies the onfire effect actually.

**Comment by KorGgenT** [comment]
you can leave out all the lines that say `"spell_class": "NONE"` because they default to that anyway.

**Comment by chaosvolt** [comment]
> heat damage already applies the onfire effect actually.

Oooh. I can see this being useful if I want to give secondary effects to distinguish fire attacks (especially since Dragonblood and Paragon of The Veil need their own variants). Might do either poison or bleed for Dragonfire given it might gain an infernal flavor (likely in exchange for less direct damage), while Paragon of The Veil's version will probably tinker with the anomaly-style fire instead of normal fire, and give some fittingly otherworldly status effect on impact.

>you can leave out all the lines that say "spell_class": "NONE" because they default to that anyway.

Ah right, doh. Thanks.

**Comment by chaosvolt** [comment]
Oh, and small question. Can spells be assigned classes without tacking a trait onto the player when they gain them? I'm still pondering whether or not to make Dragonblood and Paragon of The Veil spells 100% incompatible or whether to retain the limited compatibility. If that isn't an option, then mutual cancellation of their mutations would work too, and I could even re-use the current way that PoTV mutations basically subsume their counterparts from other spell categories (effectively making Paragon of The Veil take priority over Dragonblood).

**Comment by KorGgenT** [comment]
no, if a spell has an assigned class and you gain it by learning it from a spellbook, then you gain the class. if you gain the spell in another way, like a profession, then it doesn't automatically give you the trait. same for mutations that specifically teach you spells.

**Comment by chaosvolt** [comment]
Bit belatedly, thanks. I'll figure out how I'll balance those when it comes to that, then. In the meantime, I've put together the first few hints of the Dragonblood category's spell selection. It shows a general trend of being more powerful at first but less effective than comparable spellcraft spells when maxed out. 

Revenant Blood is effectively a mixture of the Revenant Crown's effect and Clarity of Blood, replacing Dragonblood's former use of Clarity of Blood as-is. Meanwhile Infernal Flame is rather nasty as a damaging spell (stacking on-fire and bleed) but the choice of field it can create will very quickly teach the player to be careful about using it. Their version of Surge of Adrenaline meanwhile isn't much different from the version you can get in Urban Awakening, as my idea here is to make the biggest differences in effect be between spellcraft spells and category-specific ones.

I'm pondering what else I could do for Dragonblood spells that have a spellcraft equivalent. Flame Armor would logically warrant some form of drawback that fits the theme and game balance of Dragonblood (having innate resistance to fire damage means it can capitalize on the spell more effectively, but it should be discouraged against spamming it somehow). Likewise, I have no idea at all what to do to differentiate their version of Might. Serpentine Shield at least I'm thinking I could add other side effects like nether attention or maybe the risk of generating dimensional fatigue on activation, but there's no immediate idea that comes to mind for what advantage it could gain (as stacking AEP_SNAKES does nothing as far as I recall). Maybe some other darkness-based debuff effect...

**Pull request merged**